### DISCUSSION TITLE
How do I profile the performance of Bevy app logic?

### DISCUSSION THREAD

**Discussion by phuocthanhdo** [bevyengine/bevy#6715] [Q&A] [answered]
Hi,

As we are going to optimization steps of our Game, we need some diagnostic & monitoring tools for system performance, especially on the bevy ECS systems.

The app is running with 550+ systems (still growing), mostly run with parallel executor. And we want to control the fps, by logging any systems running with low speed. We found 2 ways of system monitoring:

1. Use an attribute macro and mark on any potential slow system like `#[perf_trace(warn = 60, debug = 0)]`. The macro is just about getting time diff between start and end of a system. This solution doesn't require any bevy modification, but it's not a global monitor.
2. Make a fork on bevy_ecs, and add the measurement code (execution time of function) on `executor.rs` and `executor_parallel.rs`

So I'm wondering there is another better approach for this case?
Thank in advance.

**Comment by alice-i-cecile** [comment, marked as answer]
Flamegraphs produced by Tracy are generally going to be most useful here. You can see when you have inadequate parallelism, and which systems are longest.

Follow the docs on profiling Bevy games [here](https://github.com/bevyengine/bevy/blob/main/docs/profiling.md) to get started.

**Reply by phuocthanhdo** [reply]
Interesting. Will take a look. Thank you.
